Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,270 @@
+}
+
+extension CachedFramework: Decodable {
+	static func decode(_ j: JSON) -> Decoded<CachedFramework> {
+		return curry(self.init)
+			<^> j <| CachedFramework.nameKey
+			<*> j <| CachedFramework.md5Key
+	}
+}
+
+struct VersionFile {
+	let commitish: String
+	let xcodeVersion: String
+	
+	let macOS: [CachedFramework]?
+	let iOS: [CachedFramework]?
+	let watchOS: [CachedFramework]?
+	let tvOS: [CachedFramework]?
+	
+	static let commitishKey = "commitish"
+	static let xcodeVersionKey = "xcodeVersion"
+
+	subscript(_ platform: Platform) -> [CachedFramework]? {
+		switch platform {
+		case .macOS:
+			return macOS
+		case .iOS:
+			return iOS
+		case .watchOS:
+			return watchOS
+		case .tvOS:
+			return tvOS
+		}
+	}
+	
+	func toJSONObject() -> Any {
+		var dict: [String: Any] = [
+			VersionFile.commitishKey : commitish,
+			VersionFile.xcodeVersionKey : xcodeVersion
+		]
+		for platform in Platform.supportedPlatforms {
+			if let caches = self[platform] {
+				dict[platform.rawValue] = caches.map { $0.toJSONObject() }
+			}
+		}
+		return dict
+	}
+	
+	init(commitish: String, xcodeVersion: String, macOS: [CachedFramework]?, iOS: [CachedFramework]?, watchOS: [CachedFramework]?, tvOS: [CachedFramework]?) {
+		self.commitish = commitish
+		self.xcodeVersion = xcodeVersion
+		
+		self.macOS = macOS
+		self.iOS = iOS
+		self.watchOS = watchOS
+		self.tvOS = tvOS
+	}
+	
+	init?(url: URL) {
+		guard FileManager.default.fileExists(atPath: url.path),
+			let jsonData = try? Data(contentsOf: url),
+			let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments),
+			let versionFile: VersionFile = Argo.decode(json) else {
+			return nil
+		}
+		self = versionFile
+	}
+	
+	func md5s(for cachedFrameworks: [CachedFramework], platform: Platform, binariesDirectoryURL: URL) -> SignalProducer<String?, CarthageError> {
+		return SignalProducer(cachedFrameworks)
+			.flatMap(.concat) { cachedFramework -> SignalProducer<String?, CarthageError> in
+				let frameworkName = cachedFramework.name
+				let frameworkBinaryURL = binariesDirectoryURL
+					.appendingPathComponent(platform.rawValue, isDirectory: true)
+					.resolvingSymlinksInPath()
+					.appendingPathComponent("\(frameworkName).framework", isDirectory: true)
+					.appendingPathComponent("\(frameworkName)", isDirectory: false)
+				return md5ForFileAtURL(frameworkBinaryURL)
+					.map { md5 -> String? in
+						return md5
+					}
+					.flatMapError { _ in
+						return SignalProducer(value: nil)
+					}
+			}
+	}
+
+	func satisfies(platform: Platform, commitish: String, xcodeVersion: String, binariesDirectoryURL: URL) -> SignalProducer<Bool, CarthageError> {
+		guard let cachedFrameworks = self[platform] else {
+			return SignalProducer(value: false)
+		}
+		
+		let md5s = self.md5s(for: cachedFrameworks, platform: platform, binariesDirectoryURL: binariesDirectoryURL)
+		return self.satisfies(platform: platform, commitish: commitish, xcodeVersion: xcodeVersion, md5s: md5s)
+	}
```

Comment:
I don't think this would work. Each dependency doesn't have a separate folder for the build products it produced, so the disk is not a source of truth for what was produced by this specific dependency. 

We currently get this list by collecting on the urls spit out by the build scheme signal, so the current createVersionFile that takes a list of urls would still need to exist in some form.

The closest I think we could come is create a method that creates a comparison only version file based on the disk state, given a reference version file (putting any frameworks matching the reference version file into their own), so we could encapsulate the comparison logic into the struct itself. But it wouldn't be able to be reused for writing out the 'real' version file